src/swig/CMakeLists.txt

```diff
@@ -71,3 +73,42 @@ if (GUILE_FOUND)
   set (GUILE_EXTENSION_PATH ${_REL_GUILE_EXTENSION_PATH})
   install (TARGETS nlopt_guile LIBRARY DESTINATION ${GUILE_EXTENSION_PATH})
 endif ()
+
+
+if (JNI_FOUND AND Java_FOUND AND SWIG_FOUND)
+
+  include (UseJava)
+
+  set (SWIG_MODULE_nlopt_java_EXTRA_DEPS nlopt-java.i generate-cpp)
+  set (CMAKE_SWIG_FLAGS -package nlopt)
+
+  # swig_add_module is deprecated
+  # OUTPUT_DIR is ${CMAKE_CURRENT_BINARY_DIR}/java/ + the -package above (with
+  # any '.' replaced by '/'). It must also match the GLOB in glob_java.cmake.
+  swig_add_library (nlopt_java LANGUAGE java SOURCES nlopt.i
+                               OUTPUT_DIR ${CMAKE_CURRENT_BINARY_DIR}/java/nlopt
+                               OUTFILE_DIR ${CMAKE_CURRENT_BINARY_DIR})
+
+  set (CMAKE_SWIG_FLAGS)
+
+  swig_link_libraries (nlopt_java ${nlopt_lib})
+  target_link_libraries (nlopt_java JNI::JNI)
+
+  set_target_properties (nlopt_java PROPERTIES OUTPUT_NAME nloptjni)
+
+  install (TARGETS nlopt_java LIBRARY DESTINATION ${NLOPT_INSTALL_LIBDIR})
+
+  # unfortunately, SWIG will not tell us which .java files it generated, so we
+  # have to find out ourselves - this is the only portable way to do so
+  # (The nlopt*.i dependencies are there to force updating the list of sources
+  # on any changes to the SWIG interface code, they are not direct inputs.)
+  add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/java_sources.txt
+                     COMMAND ${CMAKE_COMMAND}
+                             -DBINARY_DIR=${CMAKE_CURRENT_BINARY_DIR}
+                             -P ${CMAKE_CURRENT_SOURCE_DIR}/glob_java.cmake
+                     DEPENDS nlopt.i nlopt-exceptions.i nlopt-java.i
+                             nlopt_java_swig_compilation glob_java.cmake)
+
+  add_jar (nlopt_jar SOURCES @${CMAKE_CURRENT_BINARY_DIR}/java_sources.txt
+                     OUTPUT_NAME nlopt)
```

Comment:
we also need an install rule, into share/java for example

Reply:
Oops, true, I will fix that too.

Reply:
The `.java` sources need not be installed anywhere, the compiled `.class` files in the `.jar` are all that the Java compiler needs to build code depending on the library. The runtime then needs both the `.jar` and the JNI `.so`.

So only the install rule for the `.jar` is missing, I am adding that one now.

Reply:
This should be fixed in the new revision (commit 32a4a64e7b087fb0cec719ccd794ed5a8fe2f585). (I added the missing `install_jar` command.)